packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,337 @@
+ * @param props.value The current value of the input field. If a value is passed to this prop, component will be controlled by consumer.
+ * @param props.className className supplied to the TextInput container.
+ */
+const TextInput = React.forwardRef(
+  (
+    {
+      label,
+      description,
+      optional = false,
+      disabled = false,
+      onChange,
+      placeholder,
+      errorMessage,
+      state,
+      value: controlledValue,
+      className,
+    }: TextInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) => {
+    const isControlled = typeof controlledValue === 'string';
+    const [uncontrolledValue, setUncontrolledValue] = useState('');
+    const value = isControlled ? controlledValue : uncontrolledValue;
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.Error) {
+        return uiColors.red.base;
+      }
+
+      if (state === State.Valid) {
+        return uiColors.green.base;
+      }
+
+      return '#CCC';
+    }
+
+    function getInputPaddingFromState() {
+      if (state === State.Valid || state === State.Error) {
+        return '30px';
+      }
+
+      if (optional) {
+        return '60px';
+      }
+
+      return '12px';
+    }
+
+    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
+      if (onChange !== undefined) {
+        onChange(e);
+      }
+
+      if (!isControlled) {
+        setUncontrolledValue(e.target.value);
+      }
+    }
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const descriptionStyle = css`
+      color: ${uiColors.gray.dark1};
+      font-size: 14px;
+      height: ${description === '' ? '0' : '20'}px;
+      font-weight: normal;
+      padding-bottom: 8px;
+    `;
+
+    const inputStyle = css`
+      width: 400px;
+      height: 36px;
+      border-radius: 4px;
+      padding-left: 12px;
+      font-size: 14px;
+      font-weight: normal;
+      z-index: ${hasFocus ? 2 : 1};
+      background-color: ${disabled ? uiColors.gray.light2 : '#FFFFFF'};
+      padding-right: ${getInputPaddingFromState()};
+      border: 1px solid ${getInputColorFromState()};
+      &:placeholder {
+        color: ${uiColors.gray.base};
+      }
+      &:focus {
+        outline: none;
+      }
+    `;
+
+    return (
+      <label className={cx(textInputStyle, labelStyle, className)}>
+        {label}
+        <label className={descriptionStyle}>{description}</label>
+        <div className={inputContainerStyle}>
```

Comment:
do we need this div or can these styles go somewhere else?

Reply:
yes, we need the div so that the icons can be vertically centered